tools/cldr-unittest/src/org/unicode/cldr/unittest/TestPaths.java

```diff
@@ -100,62 +100,117 @@ private String getPaths(Set<PathHeader> altPaths) {
+        /*
+         * Use the pathsSeen hash to keep track of which paths have
+         * already been seen. Since the test checkPrettyPaths isn't really
+         * locale-dependent, run it only once for each path, for the first
+         * locale in which the path occurs.
+         */
+        Set<String> pathsSeen = new HashSet<String>();
+        CLDRFile englishFile = testInfo.getCldrFactory().make("en", true);
+        PathHeader.Factory phf = PathHeader.getFactory(englishFile);
         Status status = new Status();
-
         for (String locale : getLocalesToTest()) {
             CLDRFile file = testInfo.getCLDRFile(locale, true);
-            logln(locale);
-
+            logln("Testing path headers and values for locale => " + locale);
             for (Iterator<String> it = file.iterator(); it.hasNext();) {
                 String path = it.next();
-                String fullPath = file.getFullXPath(path);
-                String value = file.getStringValue(path);
-                String source = file.getSourceLocaleID(path, status);
-                if (fullPath == null) {
-                    errln("Locale: " + locale + ",\t FullPath: " + path);
-                }
-                if (value == null) {
-                    errln("Locale: " + locale + ",\t Value: " + path);
-                }
-                if (source == null) {
-                    errln("Locale: " + locale + ",\t Source: " + path);
+                checkFullpathValue(path, file, locale, status, false /* not extra path */);
+                if (!pathsSeen.contains(path)) {
+                    pathsSeen.add(path);
+                    checkPrettyPaths(path, phf);
                 }
-                if (status.pathWhereFound == null) {
-                    errln("Locale: " + locale + ",\t Found Path: " + path);
+            }
+            for (String path : file.getExtraPaths()) {
+                checkFullpathValue(path, file, locale, status, true /* extra path */);
+                if (!pathsSeen.contains(path)) {
+                    pathsSeen.add(path);
+                    checkPrettyPaths(path, phf);
                 }
             }
         }
     }
 
-    public void TestPathHeaders() {
-
-        Set<String> pathsSeen = new HashSet<String>();
-        CLDRFile englishFile = testInfo.getCldrFactory().make("en", true);
-        PathHeader.Factory phf = PathHeader.getFactory(englishFile);
-
-        for (String locale : getLocalesToTest()) {
-            CLDRFile file = testInfo.getCldrFactory().make(locale, true);
-            logln("Testing path headers for locale => " + locale);
-
-            for (Iterator<String> it = file.iterator(); it.hasNext();) {
-                checkPaths(it.next(), pathsSeen, phf, locale);
-            }
-            for (String path : file.getExtraPaths()) {
-                checkPaths(path, pathsSeen, phf, locale);
+    /**
+     * For the given path and CLDRFile, check that fullPath, value, and source are all non-null.
```

Comment:
Add a comment to the code that generates the extra paths, to make sure that updates there are reflected here and vice versa.

Reply:
OK, I added two comments:
`
     * NOTE: values may be null for some "extra" paths in locales for which no explicit
     * values have been submitted. Both unit tests and Survey Tool client code generate
     * errors or warnings for null value, but allow null value for certain exceptional
     * extra paths. See the functions named extraPathAllowsNullValue in TestPaths.java
     * and in the JavaScript client code. Make sure that updates here are reflected there
     * and vice versa.
`
and
` 
     * Extra paths are generated by CLDRFile.getRawExtraPathsPrivate; this function may need
     * updating (to allow null for other paths) if that function changes.
     *
`